akka-typed/src/test/scala/docs/akka/typed/IntroSpec2.scala

```diff
@@ -0,0 +1,144 @@
+    final case class SessionDenied(reason: String) extends SessionEvent
+    final case class MessagePosted(screenName: String, message: String) extends SessionEvent
+
+    final case class PostMessage(message: String)
+    //#chatroom-protocol
+    //#chatroom-behavior
+
+    val behavior: Behavior[GetSession] = {
+      var sessions = List.empty[ActorRef[SessionEvent]]
+
+      // TODO it's not obvious that this type annotation is needed, is there a better way?
+      val b: Behavior[Command] =
+        Stateful {
+          case (ctx, GetSession(screenName, client)) ⇒
+            sessions ::= client
+            val wrapper = ctx.spawnAdapter {
+              p: PostMessage ⇒ PostSessionMessage(screenName, p.message)
+            }
+            client ! SessionGranted(wrapper)
+            Same
+          case (ctx, PostSessionMessage(screenName, message)) ⇒
+            val mp = MessagePosted(screenName, message)
+            sessions foreach (_ ! mp)
+            Same
+        }
+      b.narrow // only expose GetSession to the outside
+      //#chatroom-behavior
+    }
+  }
+  //#chatroom-actor
+
+}
+
+class IntroSpec2 extends TypedSpec {
+  import IntroSpec2._
+
+  def `must say hello`(): Unit = {
+    //#hello-world
+    import HelloWorld._
+    // TODO this is not something we would like to encourage and have in docs
+    // using global pool since we want to run tasks after system.terminate
+    import scala.concurrent.ExecutionContext.Implicits.global
+
+    val system: ActorSystem[Greet] = ActorSystem("hello", greeter)
+
+    val future: Future[Greeted] = system ? (Greet("world", _))
+
+    for {
+      greeting ← future.recover { case ex ⇒ ex.getMessage }
+      done ← { println(s"result: $greeting"); system.terminate() }
+    } println("system terminated")
+    //#hello-world
+  }
+
+  def `must chat`(): Unit = {
+    //#chatroom-gabbler
+    import ChatRoom._
+
+    val gabbler: Behavior[SessionEvent] =
+      Stateful { (_, msg) ⇒
+        msg match { // TODO what happens with unhandled messages when written like this?
+          case SessionDenied(reason) ⇒
+            println(s"cannot start chat room session: $reason")
+            Stopped
+          case SessionGranted(handle) ⇒
+            handle ! PostMessage("Hello World!")
+            Same
+          case MessagePosted(screenName, message) ⇒
+            println(s"message has been posted by '$screenName': $message")
+            Stopped
+        }
+      }
+    //#chatroom-gabbler
+
+    //#chatroom-main
+    val main: Behavior[akka.NotUsed] =
+      SignalOrMessage(
+        signal = {
+        case (ctx, PreStart) ⇒
+          val chatRoom = ctx.spawn(ChatRoom.behavior, "chatroom")
+          val gabblerRef = ctx.spawn(gabbler, "gabbler")
+          ctx.watch(gabblerRef)
+          chatRoom ! GetSession("ol’ Gabbler", gabblerRef)
+          Same
+        case (_, Terminated(ref)) ⇒
+          Stopped
+        case _ ⇒ // TODO why not PartialFunction?
+          Same
```

Comment:
yes, I was not up to speed with that `Unhandled` should be used. Kind of liked `PartialFunction` but I can understand the benefit of not using that (no tuple allocation).